📊📈 - OpenEBS analytics using Ga4

tiagolobocastro opened the pull request:
feat(analytics): add openebs analytics

    Adds a configmap allowing us to track installs
    Adds 24h ping
    Adds volume provision events

---

    fix(resize): set timeout to 30s
    
    Allows for a longer time spend in the resizer pod, which can happen
    as the pod is quite cpu limited by default.

---

    fix(k8s): try to load in-cluster first
    
    Removes the weird error message about the config file

---

    fix: set provisioner version from metadata

---

    fix(analytics): set the actual hostname

    This ensures the certificate validation is done with correct hostname
    and not the resolved ip address.

---

    docs: fix a few typos

tiagolobocastro committed:
docs: fix a few typos

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro committed:
fix(analytics): set the actual hostname

This ensures the certificate validation is done with correct hostname
and not the resolved ip address.

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro committed:
fix: set provisioner version from metadata

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro committed:
feat(analytics): add openebs analytics

Adds a configmap allowing us to track installs
Adds 24h ping
Adds volume provision events

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

niladrih commented on rawfile/analytics/ga4/usage.py:
```diff
@@ -1,18 +1,38 @@
+    def volume_provision(self, name, capacity):
+        self.builder.category("volume_provision").volume_name(name).volume_capacity(
+            capacity
+        ).replica_count(1)
```
On the localpvs, we're doing [this](https://github.com/openebs/google-analytics-4/blob/develop/usage/const.go#L21). We're doing that because the older library did that as well. It's a string which says 'replica:1'. There's no need to make any change to what you're doing here, just a heads up.

tiagolobocastro replied:
Example so it'd be this?
```json
....,
"vol_replica_count": "replica:1"
```

Abhinandan-Purkait replied:
No, `vol_replica_count` is supposed to be 1, there are 3 fields which are common in every custom event, `event_category`, `event_value`, `event_action`, these 3 need to be present in every custom event + whatever field we want.

So for volume creation, the data would be like this,

1. `event_action`: `replica:1`
2. `event_category`: `volume_provision`
3. `event_value`: `50` (size in iBytes without the unit, just the integer, this is because then in UI it automatically adds it up to show total provision in a day)
4. `vol_replica_count`: `1` --> Custom field for replica count
5. `vol_capacity`: `50Gi` --> Custom field for capacity

The first 3 fields are special as they help up formulate charts later.

Abhinandan-Purkait replied:
Its important that the `event_value` is always an integer,

for install and ping events its number of nodes, for provision and deprovision its the size in iBytes with only the integer part.

tiagolobocastro replied:
Discussed offline, the `event_*` are not used in the openebs analytics atm

niladrih commented on rawfile/rawfile.py:
```diff
@@ -26,6 +27,10 @@
 @click.option("--reserved-capacity", envvar="RESERVED_CAPACITY", default="0")
 @click.option("--capacity-override", envvar="CAPACITY_OVERRIDE", default="0")
+@click.option("--ga-enabled/--ga-disabled", is_flag=True)
+@click.option("--ga-id", envvar="GA_ID", default="G-LPD03CGEVN")
+@click.option("--ga-key", envvar="GA_KEY", default="ibj6b0nxRqC5ti4xzUgRow")
```
These are the values to the actual openebs property: https://github.com/openebs/google-analytics-4/blob/develop/usage/const.go#L5

Does this need to be disabled CI?

tiagolobocastro replied:
Yes I think we should disable on the CI
It's only enabled now so I can check it works, once I'm happy it's working I'll both disable it on CI and set the actual ga properties

tiagolobocastro replied:
Disabled on CI

mhkarimi1383 commented on deploy/helm/rawfile-localpv/values.yaml:
```diff
@@ -20,6 +20,8 @@ global:
   # -- Default image pull secret for images
   imagePullSecrets: []
+  analytics:
+    enabled: true
```
Please add docs for this option

tiagolobocastro replied:
Done

mhkarimi1383 commented on rawfile/analytics/ga4/__init__.py:
```diff
@@ -1,6 +1,81 @@
+        thread.start()
+
+
+def enabled() -> bool:
```
```suggestion
 enabled = lambda: bool(CONFIG["ga_enabled"]) # cast to ensures type hints
```

tiagolobocastro replied:
Done

mhkarimi1383 commented on rawfile/rawfile_servicer.py:
```diff
@@ -221,6 +222,11 @@ def CreateVolume(self, request, context):
+        def volume_provision(usage: Usage):
+            usage.volume_provision(request.name, size)
+
+        send_event(volume_provision)
```
in `send_event` calls should we ignore errors, So CSI call wont don't get error?

tiagolobocastro replied:
There should not be any errors on `send_event` since we only push to the queue?

mhkarimi1383 replied:
Oh, I have not seen that

mhkarimi1383 commented on rawfile/analytics/ga4/__init__.py:
```diff
@@ -1,6 +1,81 @@
+    stop_event.set()
+
+
+def send_event(event_fn):
```
send_event = lambda event_fn : task_queue.put(event_fn)

tiagolobocastro replied:
the linter complains about these lambda, asking me to add them as def (same thing on the enable)

mhkarimi1383 replied:
Ok, no problem

I have suggested them since they where some simple functions or aliases

tiagolobocastro replied:
Just for completeness, adding here:
```
rawfile/analytics/ga4/__init__.py:52:1: E731 Do not assign a `lambda` expression, use a `def`
   |
52 | shutdown_event_worker = lambda : stop_event.set()
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ E731
   |
   = help: Rewrite `shutdown_event_worker` as a `def`
```

mhkarimi1383 commented on rawfile/analytics/ga4/ping.py:
```diff
@@ -0,0 +1,111 @@
+        return self.install_ts
+
+
+def time_to_str(tm) -> str:
```
```suggestion
time_to_str = lambda tm : str(tm.strftime(time_fmt))
```

mhkarimi1383 commented on rawfile/analytics/ga4/ping.py:
```diff
@@ -0,0 +1,111 @@
+    return tm.strftime(time_fmt)
+
+
+def now_to_str() -> str:
```
```suggestion
now_to_str = lambda : datetime.now(timezone.utc).strftime(time_fmt)
def now_to_str() -> str:
```

mhkarimi1383 commented on rawfile/analytics/ga4/__init__.py:
```diff
@@ -1,6 +1,81 @@
+    logger.info("Event Worker Exited")
+
+
+def shutdown_event_worker():
```
```suggestion
shutdown_event_worker = lambda : stop_event.set()
def shutdown_event_worker():
```

mhkarimi1383 commented on rawfile/analytics/ga4/event.py:
```diff
@@ -1,66 +1,103 @@
+            PROVISIONER_VERSION
+        )
 
+    # Specify Project name, ex OpenEBS
```
Can you please make them to be docstring instead of comments?

tiagolobocastro committed:
fix(k8s): try to load in-cluster first

Removes the weird error message about the config file

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

Abhinandan-Purkait commented on rawfile/analytics/ga4/usage.py:
```diff
@@ -1,18 +1,54 @@
         event = self.builder.build()
-        self.client.send_event("openebs_event", event)
+        category = event["event_category"] if "event_category" in event else "unknown"
+        engine_name = event["engine_name"] if "engine_name" in event else "unknown"
```
Not sure if we should send if it's unknown without the category we can't really formulate any chart/graph, although the event count would be up.

tiagolobocastro replied:
Raising an exception now if not present

Abhinandan-Purkait commented on rawfile/analytics/ga4/usage.py:
```diff
@@ -1,18 +1,54 @@
+    def volume_provision(self, name: str, capacity: int):
+        self.builder.category("volume-provision").volume_name(name).volume_capacity(
+            capacity
+        ).replica_count(1)
```
How are the other fields set here? I couldn't find it here?

Abhinandan-Purkait replied:
Like, `event_action` and `event_value`, etc

Abhinandan-Purkait commented on rawfile/consts.py:
```diff
@@ -11,3 +14,5 @@
 VOLUME_IN_USE_EXIT_CODE = 102
 VOLUME_IS_ATTACHED = "openebs.io/rawfile/is-attached"
 FORMAT_OPTIONS_KEY = "openebs.io/rawfile/format-options"
+GA_ID = "Ry1MUEQwM0NHRVZO"
+GA_KEY = "aWJqNmIwbnhScUM1dGk0eHpVZ1Jvdw=="
```
Is this encoded?

tiagolobocastro replied:
Yes it is

tiagolobocastro committed:
fix(resize): set timeout to 30s

Allows for a longer time spend in the resizer pod, which can happen
as the pod is quite cpu limited by default.

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro commented:
@mergifyio queue

tiagolobocastro commented:
Resolves #111
